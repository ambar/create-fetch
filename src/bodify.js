import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'

const isPlainObject = obj =>
  obj !== null &&
  typeof obj === 'object' &&
  Object.getPrototypeOf(obj) === Object.prototype

/**
 * Stringify body
 */
const bodify = () => fetch => (url, {body, ...options} = {}) => {
  let headers = toHeaders(options.headers)
  const contentType = 'content-type'
  if (body instanceof URLSearchParams) {
    // patches old browsers
    if (!headers.get(contentType)) {
      headers.set(contentType, 'application/x-www-form-urlencoded')
    }
  } else if (isPlainObject(body) || Array.isArray(body)) {
    body = JSON.stringify(body)
    if (!headers.get(contentType)) {
      headers.set(contentType, 'application/json')
    }
  }

  return fetch(url, {
    ...options,
    body,
    headers: toObject(headers),
  })
}

export default bodify
