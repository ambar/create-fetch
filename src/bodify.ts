import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'
import isPlainObject from './utils/isPlainObject'
import {FetchEnhancer} from './types'

/* TODO: override body type
type JSONValue = string | number | boolean | JSONObject | JSONArray

interface JSONObject {
  [x: string]: JSONValue
}

interface JSONArray extends Array<JSONValue> {}

type BodifyInit = {
  body: JSONObject | JSONArray | BodyInit
}
*/

/**
 * Stringify body
 */
const bodify =
  (): FetchEnhancer =>
  (fetch) =>
  (url, {body, ...options} = {}) => {
    let headers = toHeaders(options.headers)
    const contentType = 'content-type'
    if (body instanceof URLSearchParams) {
      // patches old browsers
      if (!headers.get(contentType)) {
        headers.set(
          contentType,
          'application/x-www-form-urlencoded;charset=UTF-8'
        )
      }
    } else if (isPlainObject(body) || Array.isArray(body)) {
      body = JSON.stringify(body)
      if (!headers.get(contentType)) {
        headers.set(contentType, 'application/json')
      }
    }

    return fetch(url, {
      ...options,
      body,
      headers: toObject(headers),
    })
  }

export default bodify
