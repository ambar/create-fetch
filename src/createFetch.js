const compose = (...fns) => arg => fns.reduceRight((acc, f) => f(acc), arg)

const createFetch = (fetch, middlewares = []) => compose(...middlewares)(fetch)

export default createFetch
