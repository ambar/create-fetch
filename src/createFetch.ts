import {FetchEnhancer, FetchCompose, FetchCreate} from './types'

const compose =
  (...fns) =>
  (arg) =>
    fns.reduceRight((acc, f) => f(acc), arg)

export const composeFetch: FetchCompose = compose
export const createFetch: FetchCreate = (fetch, enhancers: FetchEnhancer[] = []) => compose(...enhancers)(fetch)
export default createFetch
