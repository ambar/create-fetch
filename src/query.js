/**
 * Stringify query string
 */
const query = () => fetch => (url, {query, ...options} = {}) => {
  if (query !== null && typeof query === 'object') {
    const params = (query instanceof URLSearchParams
      ? query
      : new URLSearchParams(query)
    ).toString()
    if (params) {
      const index = url.lastIndexOf('?')
      url += (index === -1 ? '?' : index === url.length - 1 ? '' : '&') + params
    }
  }

  return fetch(url, options)
}

export default query
