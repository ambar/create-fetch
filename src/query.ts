import getEntries from './utils/getEntries'
import {FetchEnhancer} from './types'

// preserve legacy querystring/qs behavior, set null or undefined to empty
const mapNilToEmpty = (obj: Record<string, any>) =>
  getEntries(obj).map(([k, v]) => [k, (v as string) ?? ''])

type QueryInit = {
  query?: string[][] | Record<string, any> | URLSearchParams
}

/**
 * Stringify query string
 */
const query =
  (): FetchEnhancer<QueryInit> =>
  (fetch) =>
  (url, {query, ...options} = {}) => {
    if (query !== null && typeof query === 'object') {
      const params = (
        query instanceof URLSearchParams
          ? query
          : new URLSearchParams(mapNilToEmpty(query))
      ).toString()
      if (params) {
        url = String(url)
        const index = url.lastIndexOf('?')
        url +=
          (index === -1 ? '?' : index === url.length - 1 ? '' : '&') + params
      }
    }

    return fetch(url, options)
  }

export default query
