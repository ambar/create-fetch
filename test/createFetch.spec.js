import createFetch from '../src/createFetch'
import query from '../src/query'
import bodify from '../src/bodify'

describe('createFetch', () => {
  const fetch = (...args) => Promise.resolve(args)

  it('should create fetch without middlewares', async () => {
    const myFetch = createFetch(fetch)
    expect(await myFetch('/api')).toMatchSnapshot()
    expect(await myFetch('/api', {headers: {'x-versoin': 1}})).toMatchSnapshot()
  })

  it('should create fetch with middlewares', async () => {
    const presets = [query(), bodify()]
    const myFetch = createFetch(fetch, presets)

    // default
    expect(await myFetch('/api')).toMatchSnapshot()
    // query string
    expect(await myFetch('/api', {query: {filter: 'name'}})).toMatchSnapshot()
    // json
    expect(
      await myFetch('/api', {method: 'POST', body: {name: 'JoJo'}})
    ).toMatchSnapshot()
    // form
    expect(
      await myFetch('/api', {
        method: 'POST',
        query: {filter: 'name'},
        body: new URLSearchParams({name: 'JoJo'}),
      })
    ).toMatchSnapshot()
  })

  it('should create fetch with compose', async () => {
    const next = onNext => fetch => (url, options) => {
      onNext(url, options)
      return fetch(url, options)
    }

    const onNext = jest.fn()
    const myFetch = createFetch(fetch, [query(), next(onNext)])
    await myFetch('/api', {method: 'POST', query: {foo: 1}})
    expect(onNext).toBeCalledWith('/api?foo=1', {method: 'POST'})
  })
})
