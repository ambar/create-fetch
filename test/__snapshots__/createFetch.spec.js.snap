// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFetch should create fetch with middlewares 1`] = `
Array [
  "/api",
  Object {
    "body": undefined,
    "headers": Object {},
  },
]
`;

exports[`createFetch should create fetch with middlewares 2`] = `
Array [
  "/api?filter=name",
  Object {
    "body": undefined,
    "headers": Object {},
  },
]
`;

exports[`createFetch should create fetch with middlewares 3`] = `
Array [
  "/api",
  Object {
    "body": "{\\"name\\":\\"JoJo\\"}",
    "headers": Object {
      "content-type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`createFetch should create fetch with middlewares 4`] = `
Array [
  "/api?filter=name",
  Object {
    "body": URLSearchParams: "name=JoJo",
    "headers": Object {
      "content-type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
  },
]
`;

exports[`createFetch should create fetch without middlewares 1`] = `
Array [
  "/api",
]
`;

exports[`createFetch should create fetch without middlewares 2`] = `
Array [
  "/api",
  Object {
    "headers": Object {
      "x-versoin": 1,
    },
  },
]
`;
